apps/meteor/client/providers/VideoConfPopupProvider.tsx

```diff
@@ -1,37 +1,74 @@
-import { VideoConfManager, useVideoConfIncomingCalls } from '../lib/VideoConfManager';
+import { VideoConfManager, useVideoConfIncomingCalls, useIsRinging } from '../lib/VideoConfManager';
 import VideoConfPopupPortal from '../portals/VideoConfPopupPortal';
 import VideoConfPopup from '../views/room/contextualBar/VideoConference/VideoConfPopup';
 
 const VideoConfContextProvider = ({ children }: { children: ReactNode }): ReactElement => {
 	const [popUps, setPopUps] = useState<VideoConfPopupPayload[]>([]);
+	const incomingCalls = useVideoConfIncomingCalls();
+	const customSound = useCustomSound();
+	const isRinging = useIsRinging();
 
 	const contextValue = useMemo(
 		() => ({
-			dispatch: (option: Omit<VideoConfPopupPayload, 'id'>): void => setPopUps((popUps) => [...popUps, { id: option.room._id, ...option }]),
-			dismiss: (rid: VideoConfPopupPayload['room']['_id']): void => setPopUps((prevState) => prevState.filter((popUp) => popUp.id !== rid)),
-			startCall: (rid: IRoom['_id'], title?: string): Promise<void> => VideoConfManager.startCall(rid, title),
+			dispatch: (option: Omit<VideoConfPopupPayload, 'id'> & { id?: string }): void =>
+				setPopUps((popUps) => [...popUps, { ...(option.id ? { id: option.id } : { id: option.rid }), ...option }]),
+			dismiss: (id: VideoConfPopupPayload['id']): void => setPopUps((prevState) => prevState.filter((popUp) => popUp.id !== id)),
+			startCall: (rid: IRoom['_id']): Promise<void> => VideoConfManager.startCall(rid),
 			acceptCall: (callId: string): void => VideoConfManager.acceptIncomingCall(callId),
 			joinCall: (callId: string): Promise<void> => VideoConfManager.joinCall(callId),
-			muteCall: (callId: string): void => VideoConfManager.muteIncomingCall(callId),
+			muteCall: (): void => VideoConfManager.muteIncomingCalls(),
+			rejectIncomingCall: (callId: string): void => VideoConfManager.rejectIncomingCall(callId),
 			abortCall: (): void => VideoConfManager.abortCall(),
 			useIncomingCalls: (): VideoConfIncomingCall[] => useVideoConfIncomingCalls(),
+			setPreferences: (prefs: Partial<typeof VideoConfManager['preferences']>): void => VideoConfManager.setPreferences(prefs),
+			changePreference: (key: 'cam' | 'mic', value: boolean): void => VideoConfManager.changePreference(key, value),
+			useIsRinging: (): boolean => useIsRinging(),
 		}),
 		[],
 	);
 
+	useEffect(() => {
+		if (incomingCalls.length) {
+			incomingCalls.map((incomingCall) => contextValue.dispatch({ id: incomingCall.callId, rid: incomingCall.rid, isReceiving: true }));
+
+			if (isRinging) {
+				customSound.play('calling');
+				const soundInterval = setInterval(() => {
+					customSound.play('calling');
+				}, 3000);
+
+				return (): void => {
+					customSound.pause('calling');
+					clearInterval(soundInterval);
+				};
+			}
+		}
+
+		return setPopUps([]);
+	}, [incomingCalls, contextValue, customSound, isRinging]);
+
```

Comment:
I think this could be an hook somehow , and not sure if `incomingCalls` is mutable or not, this could be a potencial bug.

I didnt like the `return setPopUps([]);` part as well

Reply:
solved in peer review!